docs/JAX_for_PyTorch_users.md

```diff
@@ -0,0 +1,890 @@
+- `jax.numpy` provides NumPy-like functions
+- `jax.scipy` provides SciPy-like functions
+- `jax.nn` provides common functions for neural networks: activations, softmax, one-hot encoding etc
+- `jax.lax` provides low-level XLA APIs
+- ...
+
+More details on available ops can be found in the [API reference](https://jax.readthedocs.io/en/latest/jax.html).
+
+All operations can be run on CPUs, GPUs or TPUs. By default, JAX Arrays are created on an accelerated device, while PyTorch tensors are created on CPUs.
+
+We can now try out some array operations and check for similarities between the JAX, NumPy and PyTorch APIs.
+
++++
+
+Standard NumPy-like indexing and slicing:
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+id: LAV8UBOaWHeI
+outputId: 53a67323-2375-4011-ba04-bc563f47abd1
+---
+tensor = jnp.ones((3, 4))
+print(f"First row: {tensor[0]}")
+print(f"First column: {tensor[:, 0]}")
+print(f"Last column: {tensor[..., -1]}")
+
+# Equivalent PyTorch op: tensor[:, 1] = 0
+tensor = tensor.at[:, 1].set(0)
+
+print(tensor)
+```
+
+We would like to note particular out-of-bounds indexing behaviour in JAX. In JAX the index is clamped to the bounds of the array in the indexing operations.
+
+```{code-cell} ipython3
+print(jnp.arange(10)[11])
+```
+
+Join arrays similar to `torch.cat`. Note the kwarg name: `axis` vs `dim`.
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+id: AQVP9jKcXWiR
+outputId: 9f0eb8c2-bda0-4864-9be3-f3941fcb74e8
+---
+t1 = jnp.concat([tensor, tensor, tensor], axis=1)
+print(t1)
+```
+
+Arithmetic operations. Operations below compute the matrix multiplication between two tensors. y1, y2 will have the same value.
+
+```{code-cell} ipython3
+:id: P8jcElVyYTp7
+
+# ``tensor.T`` returns the transpose of a tensor
+y1 = tensor @ tensor.T
+y2 = jnp.matmul(tensor, tensor.T)
+
+assert (y1 == y2).all()
+
+# This computes the element-wise product. z1, z2 will have the same value
+z1 = tensor * tensor
+z2 = jnp.multiply(tensor, tensor)
+
+assert (z1 == z2).all()
+```
+
+Single-element arrays. If you have a one-element array, for example by aggregating all values of a tensor into one value, you can convert it to a Python numerical value using `.item()`:
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+id: icgYp1wKYnGi
+outputId: a58d0290-145a-4cd0-d78b-10c4729c0bc7
+---
+agg = tensor.sum()
+agg_value = agg.item()
+print(agg_value, isinstance(agg_value, float), isinstance(agg, jax.Array))
+```
+
+JAX is following NumPy's lead and recommending a functional approach and thus JAX Array API exposes less array methods compared to PyTorch Tensors:
+```python
+# Commonly exposed class ops
+jax_array = jnp.ones((2, 3))
+jax_array.sum(), jax_array.mean(), jax_array.min(), jax_array.max(), jax_array.dot(jax_array), # ...
+
+tensor = torch.ones(2, 3)
+tensor.sum(), tensor.mean(), tensor.min(), tensor.max(), tensor.matmul(tensor.T), # ...
+
```

Comment:
Then break this to another cell and say "PyTorch exposes many more methods on its tensor object than either JAX or NumPy does on their respective array objects. Here are some examples of methods only available in PyTorch: